ninja-core/src/test/java/ninja/session/SessionImplTest.java

```diff
@@ -398,7 +381,101 @@ public void testThatCookieUsesContextPath() {
 
         verify(result).addCookie(cookieCaptor.capture());
         Cookie cookie = cookieCaptor.getValue();
-        Assert.assertThat(cookie.getPath(), CoreMatchers.equalTo("/my_context/"));
+        assertThat(cookie.getPath(), CoreMatchers.equalTo("/my_context/"));
+    }
+
+    @Test
+    public void testExpiryTime() {
+
+        // 1. Check that session is still saved when expiry time is set
+
+        Session sessionCookie1 = getNewSession();
+        sessionCookie1.init(context);
+
+        sessionCookie1.put("a", "2");
+
+        sessionCookie1.setExpiryTime(10 * 1000L);
+
+        assertThat(sessionCookie1.get("a"), CoreMatchers.equalTo("2"));
+
+        sessionCookie1.save(context, result);
+
+        Session sessionCookie2 = roundTrip(sessionCookie1);
+
+        assertThat(sessionCookie2.get("a"), CoreMatchers.equalTo("2"));
+
+        // 2. Check that session is invalidated when past the expiry time
+
+        // Set the current time past when it is called.
+        when(clock.currentTimeMillis()).thenReturn(System.currentTimeMillis() + 11 * 1000L);
+
+        Session sessionCookie3 = roundTrip(sessionCookie2);
+
+        assertNull(sessionCookie3.get("a"));
+    }
+
+    @Test
+    public void testExpiryTimeRoundTrip() {
+        // Round trip the session cookie with an expiry time in the future
+        // Then remove the expiration time to make sure it is still valid
+        when(
+            ninjaProperties.getInteger(NinjaConstant.sessionExpireTimeInSeconds))
+            .thenReturn(null);
+
+        Session sessionCookie1 = getNewSession();
+        sessionCookie1.init(context);
+
+        sessionCookie1.put("a", "2");
+
+        sessionCookie1.setExpiryTime(10 * 1000L);
+
+        assertThat(sessionCookie1.get("a"), CoreMatchers.equalTo("2"));
+
+        Session sessionCookie2 = roundTrip(sessionCookie1);
+
+        assertThat(sessionCookie2.get("a"), CoreMatchers.equalTo("2"));
+
+        assertThat(sessionCookie2.get(Session.EXPIRY_TIME_KEY), CoreMatchers.equalTo("10000"));
+
+        sessionCookie2.setExpiryTime(null);
+
+        Session sessionCookie3 = roundTrip(sessionCookie2);
+
+        assertNull(sessionCookie3.get(Session.EXPIRY_TIME_KEY));
+    }
+
+    private Session roundTrip(Session sessionCookie1) {
+        sessionCookie1.save(context, result);
+
+        // Get the cookie ...
+        verify(result, atLeastOnce()).addCookie(cookieCaptor.capture());
+
+        when(context.getCookie("NINJA_SESSION")).thenReturn(cookieCaptor.getValue());
+
+        // ... and roundtrip it into an new session
+        Session sessionCookie2 = getNewSession();
+        sessionCookie2.init(context);
+        return sessionCookie2;
+    }
+
+    private Session getNewSession() {
```

Comment:
Would you mind renaming the method to createNewSession()? .. get... always has the association to JavaBeans... which is not the case here :) Thanks!